app/scripts/controllers/network/network-controller.test.js

```diff
@@ -1,273 +1,4145 @@
+ * The set of networks that, when specified, create an Infura provider as
+ * opposed to a "standard" provider (one suited for a custom RPC endpoint).
  */
-async function withController(...args) {
-  const [constructorArgs, fn] = args.length === 2 ? args : [{}, args[0]];
-  const controller = new NetworkController({
-    ...defaultControllerOptions,
-    ...constructorArgs,
-  });
-  try {
-    return await fn({ controller });
-  } finally {
-    await controller.destroy();
-  }
-}
+const INFURA_NETWORKS = [
+  {
+    networkName: 'Mainnet',
+    networkType: 'mainnet',
+    chainId: '0x1',
+    networkVersion: '1',
+    ticker: 'ETH',
+  },
+  {
+    networkName: 'Goerli',
+    networkType: 'goerli',
+    chainId: '0x5',
+    networkVersion: '5',
+    ticker: 'GoerliETH',
+  },
+  {
+    networkName: 'Sepolia',
+    networkType: 'sepolia',
+    chainId: '0xaa36a7',
+    networkVersion: '11155111',
+    ticker: 'SepoliaETH',
+  },
+];
 
 /**
- * Setup Nock mocks for the block tracker.
- *
- * @param {object} options - Options.
- * @param {NETWORK_TYPES} options.networkType - The network type to mock.
- * @param {object} options.block - The mock block to return.
- * @param {Record<string, number>} options.requests - A set of methods to setup
- * mocks for, mapped to the number of times to expect each request.
+ * Handles mocking provider requests for a particular network.
  */
-function setupMockRpcBlockResponses({
-  networkType = NetworkController.defaultProviderConfig.type,
-  block = BLOCK,
-  requests = { eth_getBlockByNumber: 1, eth_blockNumber: 1 },
-} = {}) {
-  const rpcUrl =
-    networkType === NETWORK_TYPES.RPC
-      ? 'http://localhost:8545/'
-      : `https://${networkType}.infura.io/v3/${defaultControllerOptions.infuraProjectId}`;
-  const { origin, pathname } = new URL(rpcUrl);
-
-  const scope = nock(origin);
-
-  if (requests.eth_getBlockByNumber) {
-    scope
-      .post(
-        pathname,
-        (body) =>
-          body.method === 'eth_getBlockByNumber' &&
-          body.params[0] === block.number,
-      )
-      .times(requests.eth_getBlockByNumber)
-      .reply(200, () => JSON.stringify(constructSuccessfulRpcResponse(block)));
+class NetworkCommunications {
+  #infuraNetwork;
+
+  #customRpcUrl;
+
+  /**
+   * Builds an object for mocking provider requests.
+   *
+   * @param {object} options - An options bag.
+   * @param {"infura" | "custom"} options.networkClientType - Specifies the
+   * expected middleware stack that will represent the provider: "infura" for an
+   * Infura network; "custom" for a custom RPC endpoint.
+   * @param {string} options.infuraNetwork - The name of the Infura network
+   * being tested, assuming that `networkClientType` is "infura".
+   * @param {string} options.infuraProjectId - The project ID of the Infura
+   * network being tested, assuming that `networkClientType` is "infura".
+   * @param {string} options.customRpcUrl - The URL of the custom RPC endpoint,
```

Comment:
Nit: It seems that these three are optional:
```suggestion
   * @param {string} [options.infuraNetwork] - The name of the Infura network
   * being tested, assuming that `networkClientType` is "infura".
   * @param {string} [options.infuraProjectId] - The project ID of the Infura
   * network being tested, assuming that `networkClientType` is "infura".
   * @param {string} [options.customRpcUrl] - The URL of the custom RPC endpoint,
```

Reply:
Done here: [0b43e1b](https://github.com/MetaMask/metamask-extension/pull/17498/commits/0b43e1b6f7929e2cdaf35d20c13c73507eddf3b9)